compiler-core/src/erlang.rs

```diff
@@ -610,67 +614,72 @@ fn expr_segment<'a>(
+            value => print_expr(value).surround("(", ")"),
         }
-
-        // As normal
-        TypedExpr::Int { .. }
-        | TypedExpr::Float { .. }
-        | TypedExpr::Var { .. }
-        | TypedExpr::BitArray { .. } => expr(value, env),
-
-        // Wrap anything else in parentheses
-        value => expr(value, env).surround("(", ")"),
     };
 
-    let size = |expression: &'a TypedExpr, env: &mut Env<'a>| match expression {
-        TypedExpr::Int { value, .. } => {
-            let v = value.replace("_", "");
-            let v = u64::from_str(&v).unwrap_or(0);
-            Some(Document::String(format!(":{v}")))
-        }
+    let size = |expression: &'a TypedExpr,
+                print_expr: &mut dyn FnMut(&'a TypedExpr) -> Document<'a>| {
+        match expression {
+            TypedExpr::Int { value, .. } => {
+                let v = value.replace("_", "");
+                let v = u64::from_str(&v).unwrap_or(0);
+                Some(Document::String(format!(":{v}")))
+            }
 
-        _ => {
-            let inner_expr = expr(expression, env).surround("(", ")");
-            // The value of size must be a non-negative integer, we use lists:max here to ensure
-            // it is at least 0;
-            let value_guard = ":(lists:max(["
-                .to_doc()
-                .append(inner_expr)
-                .append(", 0]))")
-                .group();
-            Some(value_guard)
+            _ => {
+                let inner_expr = print_expr(expression).surround("(", ")");
+                // The value of size must be a non-negative integer, we use lists:max here to ensure
+                // it is at least 0;
+                let value_guard = ":(lists:max(["
+                    .to_doc()
+                    .append(inner_expr)
+                    .append(", 0]))")
+                    .group();
+                Some(value_guard)
+            }
         }
     };
 
     let unit = |value: &'a u8| Some(Document::String(format!("unit:{value}")));
 
     bit_array_segment(
-        document,
+        doc_value,
         options,
         size,
         unit,
         value_is_a_string_literal,
-        env,
+        &mut print_expr,
     )
 }
 
-fn bit_array_segment<'a, Value: 'a, SizeToDoc, UnitToDoc>(
-    mut document: Document<'a>,
+fn bit_array_segment<'a, Value: 'a, ValueType, ValueToDoc, SizeToDoc, UnitToDoc>(
+    mut value_to_doc: ValueToDoc,
     options: &'a [BitArrayOption<Value>],
     mut size_to_doc: SizeToDoc,
     mut unit_to_doc: UnitToDoc,
     value_is_a_string_literal: bool,
-    env: &mut Env<'a>,
+    print_value: &mut dyn FnMut(&'a ValueType) -> Document<'a>,
 ) -> Document<'a>
 where
-    SizeToDoc: FnMut(&'a Value, &mut Env<'a>) -> Option<Document<'a>>,
+    ValueToDoc: FnMut(&mut dyn FnMut(&'a ValueType) -> Document<'a>) -> Document<'a>,
+    SizeToDoc:
+        FnMut(&'a Value, &mut dyn FnMut(&'a ValueType) -> Document<'a>) -> Option<Document<'a>>,
```

Comment:
Please revert these to the previous signature and add an addition function type here rather than using dynamic dispatch for `print_value`.

Reply:
@lpil 
Unfortunately we cant remove the dynamic dispatch. Since the `ValueToDoc` & `SizeToDoc` are closures they don't let us use generics for passing the `print_value` into them. If we try what your proposing, the borrow checker gives us [E0782](https://doc.rust-lang.org/error_codes/E0782.html) at the `ValueToDoc` & `SizeToDoc` definitions. 

Im not sure how best to solve this, I'll try think about it today and come up with something. Buy any recommendations would be helpful.

Reply:
`print_value` should not be passed to them, it should continue to work as before.